src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -666,37 +938,39 @@ public void Prune()
                 {
                     try
                     {
-                        DirtyNodesCache cache = DirtyNodes;
-                        lock (cache)
+                        // Flush ahead of time so that memtable is empty which prevent stalling when writing nodes.
+                        // Note, the WriteBufferSize * WriteBufferNumber need to be more than about 20% of pruning cache
+                        // otherwise, it may not fit the whole dirty cache.
+                        // Additionally, if (WriteBufferSize * (WriteBufferNumber - 1)) is already more than 20% of pruning
+                        // cache, it is likely that there are enough space for it on most time, except for syncing maybe.
+                        _nodeStorage.Flush();
```

Comment:
Can wee flush earlier? Or in a separate async Task? Basically not to block here.

Reply:
We are flushing right before lock here, so we are not blocking block processing while its flushing the memtable.

Reply:
but could we flush parallel to `PruneCache`?

Reply:
Not recomended. It is still consume cpu cycle. If it can be done outside the lock, it should be done outside the lock.